File: AIClient/lib/tools/python_command.dart
/// Helper for the `list python` and `use python <lib> <details>` chat commands.
///
/// Keeps all library metadata, prompt construction, and code extraction logic
/// in one place so [ChatScreen] stays focused on UI concerns.
class PythonCommand {
  PythonCommand._();

  // ── Library registry ────────────────────────────────────────────────────────

  /// Maps each canonical library name to its recognised aliases (lower-cased)
  /// and a short human-readable description.
  static const _libs = <String, ({List<String> aliases, String description})>{
    'numpy': (aliases: ['numpy', 'np'], description: 'Numerical arrays & math'),
    'scipy': (
      aliases: ['scipy', 'sp'],
      description: 'Scientific computing & signal processing'
    ),
    'pandas': (aliases: ['pandas', 'pd'], description: 'Data analysis & DataFrames'),
    'matplotlib': (aliases: ['matplotlib', 'plt'], description: '2D static charts & plots'),
    'plotly': (aliases: ['plotly', 'px', 'go'], description: 'Interactive charts'),
    'astral': (aliases: ['astral'], description: 'Sunrise, sunset, moon phase & solar calculations'),
    'folium': (aliases: ['folium'], description: 'Interactive Leaflet.js maps with markers, polygons & heatmaps'),
    'shapely': (aliases: ['shapely'], description: 'Geometric operations — areas, distances & spatial intersections'),
  };

  // ── Public API ──────────────────────────────────────────────────────────────

  /// Formatted list of all supported libraries — suitable for a chat bubble.
  static String listMessage() {
    final buf = StringBuffer('Available Python libraries:\n\n');
    for (final entry in _libs.entries) {
      final canonical = entry.key;
      final aliases = entry.value.aliases
          .where((a) => a != canonical)
          .join(', ');
      final aliasStr = aliases.isNotEmpty ? ' (aliases: $aliases)' : '';
      buf.writeln('• $canonical$aliasStr — ${entry.value.description}');
    }
    buf.write(
      '\nUsage: use tool <request>\n'
      'Example: use tool plot a sine wave\n'
      'The model will choose the best library automatically.\n'
      '\nType "list tools" to show this list again.',
    );
    return buf.toString().trimRight();
  }

  /// Builds the prompt sent to the LLM for a `use tool <request>` command.
  /// Lists every available library so the model only picks supported ones.
  static String buildToolPrompt(String request) {
    final libNames = _libs.keys.join(', ');
    return '''
You are a Python code generator. Write code to accomplish the task below.

INSTALLED LIBRARIES (the ONLY third-party imports allowed): $libNames

Python standard library modules (math, datetime, json, re, statistics, etc.) are also available.

PRE-DEFINED VARIABLES (already set before your code runs — do NOT redefine them):
- user_latitude         (float)            — device GPS latitude
- user_longitude        (float)            — device GPS longitude
- user_timezone_offset  (float)            — device UTC offset in hours (e.g. -5.0 for CDT)
- user_timezone         (datetime.timezone) — ready-made tzinfo object; use directly as tzinfo= in astral/datetime calls
Use these directly whenever the task involves the user's current location or local time. Do NOT hardcode placeholder coordinates or timezones. Do NOT use zoneinfo or look up a timezone from coordinates — use user_timezone instead.

Task: $request

STRICT RULES — violating any rule will cause a runtime error:
1. Every import must come from the INSTALLED LIBRARIES list or the Python standard library. Do NOT invent or guess library names.
2. Always write every import statement explicitly at the top.
3. For simple maths use the built-in math module — do not import numpy just for basic arithmetic.
4. If the task needs a library NOT on the installed list, solve it with the standard library instead.
5. All output: assign the final answer to a variable named result (e.g. result = 42). CRITICAL EXCEPTIONS:
   - matplotlib: do NOT assign result at all — the PNG is auto-captured from the open figure. Never write result = "done", result = "Plot generated", result = "success", or any other string.
   - plotly: do NOT assign result — the figure is auto-captured from the 'fig' variable. If you must assign result manually, you MUST use fig.to_html(include_plotlyjs='cdn', full_html=True) — omitting full_html=True returns a bare <div> fragment, not valid HTML.
6. Do NOT call exit() or quit() anywhere in the code — these raise SystemExit and crash the runner.
   Do NOT use try/except blocks — they hide errors and cause NameError when variables assigned inside try are used outside it. Write straightforward code without exception handling.
7. Library choice for charts: use plotly for interactive charts; use matplotlib for static plots/images. If the request uses the word "interactive" or asks for hoverable/zoomable charts, always use plotly.
8. Do NOT make any network/HTTP requests from Python code — all outbound HTTP is blocked. This means: no pd.read_html(url), no scipy.datasets, no requests/urllib calls.
9. Do NOT call fig.write_image() — kaleido is not available.
10. Do NOT read or write local file paths — there is no real filesystem. Use in-memory objects (BytesIO, StringIO) if intermediate storage is needed.
11. datetime usage: always use 'import datetime' then 'datetime.date.today()' or 'datetime.datetime.now()'. Do NOT do 'from datetime import datetime' then call 'datetime.datetime.today()' — that is a double-reference error.

CORRECT IMPORT PATTERNS — use exactly these, no variations:
- numpy:      import numpy as np
- scipy:      from scipy import stats  /  from scipy.optimize import minimize  /  from scipy.signal import find_peaks  /  from scipy.spatial.distance import cdist
- pandas:     import pandas as pd
- matplotlib: import matplotlib.pyplot as plt  — do NOT call plt.show() (auto-captured)
- plotly:     import plotly.express as px  OR  import plotly.graph_objects as go  — assign figure to 'fig', do NOT call fig.show() (auto-captured)
- astral:     from astral import LocationInfo; from astral.sun import sun; from astral.moon import phase
  LocationInfo: LocationInfo(name="Place", region="", timezone="UTC", latitude=user_latitude, longitude=user_longitude)  — always pass timezone="UTC" when using user_timezone as tzinfo; kwargs are 'latitude'/'longitude', NOT 'lat'/'lon'
  sun() usage: s = sun(location.observer, date=some_date, tzinfo=user_timezone)  — first arg is POSITIONAL (location.observer), never write observer=location.observer or location.observer=anything; s is a DICT; access as s['sunrise'], s['sunset'], s['noon'], s['dawn'], s['dusk']  — do NOT call sun.sunrise() or sun.sunset(), those do not exist; do NOT use zoneinfo or look up timezone from coordinates
  phase() usage: phase(date) returns a float 0–28 (moon age in days)
  assign formatted text to result
- folium:     import folium  — create map as m = folium.Map(location=[lat,lon], zoom_start=n)  — assign result = m.get_root().render() at the end (NOT m._repr_html_() — that returns a div fragment, not a full HTML document)
- shapely:    from shapely.geometry import Point, Polygon, LineString, MultiPolygon  /  from shapely.ops import unary_union
- distance:   use the math module only — haversine formula for GPS distance: import math; def haversine(lat1,lon1,lat2,lon2): R=6371; p1,p2=math.radians(lat1),math.radians(lat2); dp=math.radians(lat2-lat1); dl=math.radians(lon2-lon1); a=math.sin(dp/2)**2+math.cos(p1)*math.cos(p2)*math.sin(dl/2)**2; return 2*R*math.asin(math.sqrt(a))

Return ONLY the Python code block, no explanation.
'''.trim();
  }

  /// Extracts Python source code from an LLM response.
  ///
  /// Tries, in order:
  /// 1. Content inside ```python … ``` fences.
  /// 2. Content inside ``` … ``` fences (any language tag).
  /// 3. The raw response, trimmed (fallback).
  ///
  /// Returns `null` only when the response is empty.
  static final _pyFenceRe =
      RegExp(r'```python\s*\n([\s\S]*?)```', caseSensitive: false);
  static final _genericFenceRe = RegExp(r'```[^\n]*\n([\s\S]*?)```');

  static String? extractCode(String llmResponse) {
    final trimmed = llmResponse.trim();
    if (trimmed.isEmpty) return null;

    // 1. ```python … ```
    final pyMatch = _pyFenceRe.firstMatch(trimmed);
    if (pyMatch != null) return pyMatch.group(1)!.trim();

    // 2. ``` … ```
    final gMatch = _genericFenceRe.firstMatch(trimmed);
    if (gMatch != null) return gMatch.group(1)!.trim();

    // 3. Raw fallback
    return trimmed;
  }
}

File: AIAndroid/app/src/main/kotlin/ai/ecoinference/app/services/PythonCommand.kt
package ai.ecoinference.app.services

/**
 * Helpers for the `list tools` and `use tool <request>` chat commands.
 * Mirrors iOS's PythonCommand.swift — library metadata, prompt construction,
 * and code extraction in one place. Ported 2026-07-28; iOS had no Android
 * equivalent of this command at all until then.
 */
object PythonCommand {

    // ── Library registry ─────────────────────────────────────────────────────

    private data class LibInfo(val aliases: List<String>, val description: String)

    private val libs: List<Pair<String, LibInfo>> = listOf(
        "numpy"      to LibInfo(listOf("numpy", "np"),        "Numerical arrays & math"),
        "scipy"      to LibInfo(listOf("scipy", "sp"),        "Scientific computing & signal processing"),
        "pandas"     to LibInfo(listOf("pandas", "pd"),       "Data analysis & DataFrames"),
        "matplotlib" to LibInfo(listOf("matplotlib", "plt"),  "2D static charts & plots"),
        "plotly"     to LibInfo(listOf("plotly", "px", "go"), "Interactive charts"),
        "astral"     to LibInfo(listOf("astral"),             "Sunrise, sunset, moon phase & solar calculations"),
        "folium"     to LibInfo(listOf("folium"),             "Interactive Leaflet.js maps with markers, polygons & heatmaps"),
        "shapely"    to LibInfo(listOf("shapely"),            "Geometric operations — areas, distances & spatial intersections"),
    )

    // ── Public API ────────────────────────────────────────────────────────────

    /** Formatted list of supported libraries for display in the chat. */
    fun listMessage(): String {
        val buf = StringBuilder("Available Python libraries:\n\n")
        for ((name, info) in libs) {
            val otherAliases = info.aliases.filter { it != name }.joinToString(", ")
            val aliasSuffix  = if (otherAliases.isEmpty()) "" else " (aliases: $otherAliases)"
            buf.append("• $name$aliasSuffix — ${info.description}\n")
        }
        buf.append("\nUsage: use tool <request>\n")
        buf.append("Example: use tool plot a sine wave\n")
        buf.append("The model will choose the best library automatically.\n")
        buf.append("\nType \"list tools\" to show this list again.")
        return buf.toString()
    }

    /**
     * Builds the LLM prompt for a `use tool <request>` command.
     *
     * Kept deliberately compact — this instructional text and the model's
     * response share one fixed total token budget (model's maxContextTokens),
     * so every character spent here is a character the generated code
     * doesn't get. Also explicitly asks for compact, non-human-readable code
     * (short names, no comments/blank lines) — Python runs identically
     * either way, so this trims the response's own token cost for free.
     * Mirrors iOS's PythonCommand.buildToolPrompt() text exactly.
     */
    fun buildToolPrompt(request: String, locationPreamble: String? = null): String {
        val libNames = libs.joinToString(",") { it.first }
        val locSection = if (locationPreamble != null)
            " Pre-defined(don't redefine):user_latitude,user_longitude(floats),user_timezone_offset(float,UTC hrs),user_timezone(tzinfo)—use for location/time tasks,never hardcode,never zoneinfo."
        else ""

        return """
Python code generator. Task:$request
Write MINIMAL code:short names,no comments/blank lines/docstrings,semicolons where natural. Correct>readable. Prefer one-liners/indexing over if-chains; if you DO write if/for/def, indent the body 4 spaces on its own line—never leave it flush left.
Only import:$libNames,or stdlib. Never guess a library.$locSection
Rules:scalar-only math→use math not numpy(numpy is for arrays—always use numpy for array/vectorized math). result=<answer> EXCEPT matplotlib(never set result,PNG auto-captured) and plotly(never set result,fig auto-captured;else fig.to_html(include_plotlyjs='cdn',full_html=True)). No exit/quit,no try/except,no network,no file I/O(use BytesIO/StringIO),no fig.write_image(). plotly=interactive,matplotlib=static. datetime:import datetime;datetime.date.today() not from datetime import datetime.
Imports:numpy as np|from scipy import stats/from scipy.optimize import minimize|pandas as pd|matplotlib.pyplot as plt(no plt.show())|plotly.express as px/plotly.graph_objects as go(assign fig,no fig.show())|astral:from astral import LocationInfo;from astral.sun import sun;from astral.moon import phase—LocationInfo(name="P",region="",timezone="UTC",latitude=user_latitude,longitude=user_longitude);sun(location.observer,date=d,tzinfo=user_timezone)→s['sunrise']/s['sunset']/s['noon'];phase(d)→float 0-27.99,date arg ONLY(no location),a NUMBER not an object;name it EXACTLY:import bisect;result=["New Moon","Waxing Crescent","First Quarter","Waxing Gibbous","Full Moon","Waning Gibbous","Last Quarter","Waning Crescent","New Moon"][bisect.bisect([1.75,5.25,8.75,12.25,15.75,19.25,22.75,26.25],phase(d))]|folium:import folium;m=folium.Map(location=[lat,lon],zoom_start=n);result=m.get_root().render()|shapely:from shapely.geometry import Point,Polygon,LineString
Return ONLY code,no explanation,no fence.
""".trimIndent()
    }

    // ── Code extraction ──────────────────────────────────────────────────────

    /**
     * Extracts Python source from an LLM response. Tries ```python fences,
     * then generic fences, then strips an unmatched single marker at either
     * end, then falls back to the raw trimmed text.
     */
    fun extractCode(response: String): String? {
        val trimmed = response.trim()
        if (trimmed.isEmpty()) return null

        // 1. ```python ... ```
        val pyFenceIdx = trimmed.indexOf("```python", ignoreCase = true)
        if (pyFenceIdx >= 0) {
            val openEnd = pyFenceIdx + "```python".length
            val closeIdx = trimmed.indexOf("```", openEnd)
            if (closeIdx >= 0) {
                val code = trimmed.substring(openEnd, closeIdx).trim()
                if (code.isNotEmpty()) return code
            }
        }

        // 2. ``` ... ```
        val openIdx = trimmed.indexOf("```")
        if (openIdx >= 0) {
            val closeIdx = trimmed.indexOf("```", openIdx + 3)
            if (closeIdx >= 0) {
                var content = trimmed.substring(openIdx + 3, closeIdx)
                val nl = content.indexOf('\n')
                if (nl >= 0) content = content.substring(nl + 1)
                val code = content.trim()
                if (code.isNotEmpty()) return code
            }
        }

        // 3. No matched fence pair. Two real cases seen in practice on iOS
        // (confirmed live, 2026-07-28), equally possible here:
        // (a) the response was cut off before the model wrote a closing ```
        //     (leading marker present, no trailing one), or
        // (b) the model started straight into code with no opening fence at
        //     all, but still tacked on a trailing ``` out of habit (trailing
        //     marker present, no leading one).
        // Either way, leaving an unmatched fence marker embedded in the
        // "code" guarantees a Python syntax error, so strip whichever single
        // marker is actually present at the start and/or end — independently,
        // since only one of the two may be there.
        var text = trimmed
        val anchoredPy = text.indexOf("```python", ignoreCase = true)
        if (anchoredPy == 0) {
            text = text.substring("```python".length)
        } else if (text.startsWith("```")) {
            text = text.substring(3)
            val nl = text.indexOf('\n')
            if (nl >= 0) text = text.substring(nl + 1)
        }
        text = text.trim()
        if (text.endsWith("```")) {
            text = text.substring(0, text.length - 3).trim()
        }

        return text.ifEmpty { null }
    }
}
